pkg/apis/nmstate/v1alpha1/condition_types.go

```diff
@@ -0,0 +1,70 @@
+package v1alpha1
+
+import (
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type ConditionList []Condition
+
+// +k8s:openapi-gen=true
+type Condition struct {
+	Type               ConditionType          `json:"type"`
+	Status             corev1.ConditionStatus `json:"status"`
+	Reason             ConditionReason        `json:"reason,omitempty"`
+	Message            string                 `json:"message,omitempty"`
+	LastHeartbeatTime  metav1.Time            `json:"lastHearbeatTime,omitempty"`
+	LastTransitionTime metav1.Time            `json:"lastTransitionTime,omitempty"`
+}
+
+// +k8s:openapi-gen=true
+type ConditionType string
+
+// +k8s:openapi-gen=true
+type ConditionReason string
+
+func NewCondition(conditionType ConditionType, status corev1.ConditionStatus, reason ConditionReason, message string) Condition {
+	now := metav1.Time{Time: time.Now()}
+	condition := Condition{
+		Type:               conditionType,
+		Status:             status,
+		Reason:             reason,
+		Message:            message,
+		LastHeartbeatTime:  now,
+		LastTransitionTime: now,
+	}
+	return condition
+}
+
+func (conditions *ConditionList) Set(conditionType ConditionType, status corev1.ConditionStatus, reason ConditionReason, message string) {
```

Comment:
I'd try to rethink the `Set` name. 
It feels like it's changing a property of the list, while what we are doing here is to update / insert an element into the set.
I'd call it `InsertForType` or something along that line. Now that I am writing this I am also thinking: why can't this be a map?
It would make a lot more clear that we are storing just one `Condition` per type.

Reply:
+1 for map.

Reply:
I would agree with you, but it is against Kubernetes API guidelines. It is always preferred to have:

```yaml
- name: foo
  val: bar
```

instad of:

```yaml
- foo:
    val: bar
```

Moreover, in this specific case of conditions, Kubernetes expects to get a List. Otherwise our conditions wound not be usable with `kubectl` and other tools.

Reply:
I don't like the `InserForType`, it does not mean anything to me. Maybe we can make it `Update`? I'm personally fine with `Set`, but open for discussion.

Reply:
If we don't go for maps Set is ok.